### PR TITLE
:sparkles: Add atomic policy

### PR DESCRIPTION
**Pull request by elbeno**
Problem:
- Microcontroller platforms may offer rudimentary atomic instructions.
- Performing a lock-free operation may be more expensive than operating under a "mutex" (e.g. if interrupts are very cheap to turn on/off).
- We want maximum control over how atomic operations are implemented to achieve best peformance.

Solution:
- Add an atomic policy which allows customization of how operations are implemented.

Notes:
- The "standard" interface is implemented with atomic builtins which lie underneath std::atomic. See https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html
- The minimal atomic interface is provided:
  - `load` and `store`
  - `exchange` (to support simple boolean latches)
  - `fetch_add` and `fetch_sub` (to support inc/dec of an atomic counter)
  - `fetch_and`, `fetch_or` and `fetch_xor` (to support bitsets)

### CODE REVIEW

File: include/conc/concurrency.hpp [hunk 1]
```diff
@@ -35,12 +35,17 @@ template <typename Mutex = std::mutex> class standard_policy {
             }
         }
     }
+
+    template <typename T>
+    constexpr static auto atomic_alignment_of() -> std::size_t {
+        return 0;
```
Comment (lukevalenty) [inline include/conc/concurrency.hpp:41]: question: why not return `alignof(T)`?

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -8,31 +8,40 @@ add_unit_test(
...
+add_compile_fail_test(fail_no_conc_policy.cpp LIBRARIES concurrency)
+
+add_unit_test(
+    atomic_standard_policy_test
+    CATCH2
+    FILES
+    atomic_standard_policy.cpp
+    LIBRARIES
+    warnings
+    concurrency)
```
Comment (lukevalenty) [inline test/CMakeLists.txt:47]: question: why not add `sanitize` here?

File: include/conc/concurrency.hpp [hunk 3]
```diff
@@ -35,12 +35,17 @@ template <typename Mutex = std::mutex> class standard_policy {
             }
         }
     }
+
+    template <typename T>
+    constexpr static auto atomic_alignment_of() -> std::size_t {
```
Comment (lukevalenty) [inline include/conc/concurrency.hpp:40]: issue: i see this is never used anywhere

**Review by lukevalenty — APPROVED** [review]

File: test/CMakeLists.txt (hunk 2 above)
Comment (elbeno) [reply test/CMakeLists.txt:47]: It's already added by `add_unit_test` based on the SANITIZERS environment variable.

File: include/conc/concurrency.hpp (hunk 3 above)
Comment (elbeno) [reply include/conc/concurrency.hpp]: It was a leftover experiment - removed.

File: include/conc/concurrency.hpp (hunk 1 above)
Comment (elbeno) [reply include/conc/concurrency.hpp]: Experiments. Gone now.

**Pull request merged**